src/main/webapp/app/course/competencies/competency-accordion/competency-accordion.component.ts

```diff
@@ -0,0 +1,167 @@
+        if (changes.openedIndex && this.index !== this.openedIndex) {
+            this.open = false;
+        }
+        if (changes.metrics) {
+            this.setNextExercises();
+            this.setNextLessonUnits();
+        }
+    }
+
+    setNextExercises() {
+        if (!this.metrics) {
+            this.nextExercises = [];
+        }
+
+        const submittedExercises = Object.keys(this.metrics.exerciseMetrics?.latestSubmission ?? {}).map(Number);
+        const competencyExercises = this.metrics.competencyMetrics?.exercises?.[this.competency.id] ?? [];
+        const nextExerciseInformations = competencyExercises
+            .filter((exerciseId) => !submittedExercises.includes(exerciseId))
+            .flatMap((exerciseId) => this.metrics.exerciseMetrics?.exerciseInformation?.[exerciseId] ?? [])
+            .filter((exercise) => exercise.startDate.isBefore(dayjs()) && exercise.dueDate.isAfter(dayjs()))
+            .sort((a, b) => a.dueDate.diff(b.dueDate));
+
+        // Workaround to convert ExerciseInformation to Exercise
+        this.nextExercises = nextExerciseInformations.map(
+            (exercise) =>
+                ({
+                    ...exercise,
+                    studentAssignedTeamIdComputed: exercise.studentAssignedTeamId,
+                }) as unknown as Exercise,
+        );
+    }
+
+    setNextLessonUnits() {
+        if (!this.metrics) {
+            this.nextLectureUnits = [];
+        }
+
+        const completedLectureUnits = this.metrics.lectureUnitStudentMetricsDTO?.completed ?? [];
+        const competencyLectureUnits = this.metrics.competencyMetrics?.lectureUnits?.[this.competency.id] ?? [];
+        this.nextLectureUnits = competencyLectureUnits
+            .filter((lectureUnitId) => !completedLectureUnits.includes(lectureUnitId))
+            .flatMap((lectureUnitId) => this.metrics.lectureUnitStudentMetricsDTO?.lectureUnitInformation?.[lectureUnitId] ?? [])
+            .filter((lectureUnit) => dayjs(lectureUnit.releaseDate).isBefore(dayjs()))
+            .sort((a, b) => dayjs(a.releaseDate).diff(dayjs(b.releaseDate)));
+    }
+
+    toggle() {
+        this.open = !this.open;
+        this.accordionToggle.emit({ opened: this.open, index: this.index });
+    }
+
+    get jolRating() {
+        return this.metrics.competencyMetrics?.jolValues?.[this.competency.id];
+    }
+
+    onRatingChange(newRating: number) {
+        if (this.metrics.competencyMetrics) {
+            this.metrics.competencyMetrics.jolValues = {
+                ...this.metrics.competencyMetrics.jolValues,
+                [this.competency.id]: newRating,
+            };
+        }
+    }
+
+    getUserProgress(): CompetencyProgress {
+        const progress = this.metrics.competencyMetrics?.progress?.[this.competency.id] ?? 0;
+        const confidence = this.metrics.competencyMetrics?.confidence?.[this.competency.id] ?? 0;
+        return { progress, confidence } as CompetencyProgress;
```

Comment:
using the `as` keyword seems unnecessary here since we already specify the return type.
```suggestion
    getUserProgress(): CompetencyProgress {
        const progress = this.metrics.competencyMetrics?.progress?.[this.competency.id] ?? 0;
        const confidence = this.metrics.competencyMetrics?.confidence?.[this.competency.id] ?? 0;
        return { progress, confidence };
```